ai-feedback/helpers/template_utils.R

```diff
@@ -0,0 +1,187 @@
+    if (has_submission_image && has_solution_image) {
+      template_data$submission_image <- "The first attached image is the student's submission."
+    } else if (has_submission_image) {
+      template_data$submission_image <- "The attached image is the student's submission."
+    } else {
+      template_data$submission_image <- "[Submission Image Attached]"
+    }
+  }
+  
+  if (grepl("\\{solution_image\\}", prompt_content) && !("solution_image" %in% names(template_data))) {
+    if (has_submission_image && has_solution_image) {
+      template_data$solution_image <- "The second attached image is the expected solution."
+    } else if (has_solution_image) {
+      template_data$solution_image <- "The attached image is the expected solution."
+    } else {
+      template_data$solution_image <- "[Solution Image Attached]"
+    }
+  }
+  
+  # Replace placeholders
+  result <- prompt_content
+  for (key in names(template_data)) {
+    placeholder <- paste0("{", key, "}")
+    if (grepl(placeholder, result, fixed = TRUE)) {
+      result <- gsub(placeholder, template_data[[key]], result, fixed = TRUE)
+    }
+  }
+  
+  remaining_placeholders <- regmatches(result, gregexpr("\\{[a-zA-Z_][a-zA-Z0-9_]*\\}", result))[[1]]
+  if (length(remaining_placeholders) > 0) {
+    stop("Missing placeholders in template: ", paste(remaining_placeholders, collapse = ", "))
+  }
+  
+  return(result)
+}
+
+#' Generate file reference descriptions
+#' 
+#' @param submission Path to student's submission file
+#' @param solution Path to instructor's solution file (optional)
+#' @param test_output Path to test output file (optional)
+#' 
+#' @return Character string with file descriptions
+gather_file_references <- function(submission, solution = NULL, test_output = NULL) {
+  references <- c()
+  
+  if (!is.null(submission)) {
+    references <- c(references, paste("The student's submission file is", basename(submission), "."))
+  }
+  
+  if (!is.null(solution)) {
+    references <- c(references, paste("The instructor's solution file is", basename(solution), "."))
+  }
+  
+  if (!is.null(test_output)) {
+    references <- c(references, paste("The student's test output file is", basename(test_output), "."))
+  }
+  
+  return(paste(references, collapse = "\n"))
+}
+
+#' Generate file contents with line numbers
+#'
+#' @param file_paths List of file paths to process
+#' 
+#' @return Character string with file contents and line numbers
+gather_file_contents <- function(file_paths) {
+  file_contents <- ""
+  
+  for (file_path in file_paths) {
+    if (is.null(file_path) || !file.exists(file_path)) {
+      next
+    }
+    
+    filename <- basename(file_path)
+    
+    tryCatch({
+      if (grepl("\\.pdf$", filename, ignore.case = TRUE)) {
+        # Handle PDF files
+        text_content <- extract_pdf_text(file_path)
+        file_contents <- paste0(file_contents, "=== ", filename, " ===\n")
```

Comment:
There's a lot of duplicated code in these two branches. The part that is different is the definition of `lines`; everything else is in common and can go after the if statement.

I'm not sure if there's the same issue in the Python version, but if there is please make a PR to fix it there as well.